src/data/W/constructions.lean

```diff
@@ -0,0 +1,144 @@
+@[simp] def to_nat : W_type nat_β → ℕ
+| (W_type.mk nat_α.zero f) := 0
+| (W_type.mk nat_α.succ f) := (to_nat (f ())).succ
+
+lemma left_inv_nat : function.left_inverse nat_to to_nat
+| (W_type.mk nat_α.zero f) := by { simp, tidy }
+| (W_type.mk nat_α.succ f) := by { simp, tidy }
+
+lemma right_inv_nat : function.right_inverse nat_to to_nat
+| nat.zero := rfl
+| (nat.succ n) := by rw [nat_to, to_nat, right_inv_nat n]
+
+/-- The naturals are equivalent to their associated `W_type` -/
+def equiv_nat : W_type nat_β ≃ ℕ :=
+{ to_fun := to_nat,
+  inv_fun := nat_to,
+  left_inv := left_inv_nat,
+  right_inv := right_inv_nat }
+
+/--
+`nat_α` is equivalent to `bool`.
+This is useful when considering the associated polynomial endofunctor
+-/
+@[simps] def nat_α_equiv_bool : nat_α ≃ bool :=
+{ to_fun := λ c, match c with | nat_α.zero := ff | nat_α.succ := tt end,
+  inv_fun := λ b, match b with | bool.tt := nat_α.succ | bool.ff := nat_α.zero end,
+  left_inv := λ c, match c with | nat_α.zero := rfl | nat_α.succ := rfl end,
+  right_inv := λ b, match b with | bool.tt := rfl | bool.ff := rfl end }
+
+end nat
+
+section list
+
+variable (γ : Type u)
+
+/--
+The constructors for lists.
+There is "one constructor `cons x` for each `x : γ`",
+since we view `list γ` as
+| nil : list γ
+| cons x₀ : list γ → list γ
+| cons x₁ : list γ → list γ
+|   ⋮      γ many times
+-/
+inductive list_α : Type u
+| nil : list_α
+| cons : γ → list_α
+
+instance : inhabited (list_α γ) := ⟨ list_α.nil ⟩
+
+/-- The arities of each constructor for lists, `nil` takes no arguments, `cons hd` takes one -/
+def list_β : list_α γ → Type u
+| list_α.nil := pempty
+| (list_α.cons hd) := punit
+
+instance (hd : γ) : inhabited (list_β γ (list_α.cons hd)) := ⟨ punit.star ⟩
+
+/-- The isomorphism from lists to the `W_type` construction of lists -/
+@[simp] def list_to : list γ → W_type (list_β γ)
+| list.nil := ⟨ list_α.nil, pempty.elim ⟩
+| (list.cons hd tl) := ⟨ list_α.cons hd, λ _ , list_to tl ⟩
+
+/-- The isomorphism from the `W_type` construction of lists to lists -/
+@[simp] def to_list : W_type (list_β γ) → list γ
+| (W_type.mk list_α.nil f) := []
+| (W_type.mk (list_α.cons hd) f) := hd :: to_list (f punit.star)
+
+lemma left_inv_list : function.left_inverse (list_to γ) (to_list _)
+| (W_type.mk list_α.nil f) := by { simp, tidy }
+| (W_type.mk (list_α.cons x) f) := by { simp, tidy }
+
+lemma right_inv_list : function.right_inverse (list_to γ) (to_list _)
+| list.nil := rfl
+| (list.cons hd tl) := by simp [right_inv_list tl]
+
+/-- Lists are equivalent to their associated `W_type` -/
+def equiv_list : W_type (list_β γ) ≃ list γ :=
+{ to_fun := to_list _,
+  inv_fun := list_to _,
+  left_inv := left_inv_list _,
+  right_inv := right_inv_list _ }
+
+/--
+`list_α` is equivalent to `γ` with an extra point.
+This is useful when considering the polynomial endofunctor associated to the `W_type`
+-/
+def list_α_equiv_punit_sum : list_α γ ≃ punit.{v + 1} ⊕ γ :=
```

Comment:
Could also consider
```suggestion
def list_α_equiv_option : list_α γ ≃ option γ :=
```
which feels simpler, but is maybe less convenient for you.

Reply:
yeah option is a possibility, though I want to make addition of monomials (in polynomial endofunctors) in general, so having a direct sum is quite nice. I might change `bool` to `punit + punit` since that is exactly what I need on the other end

Reply:
or I could make all four of the above

Reply:
Note we have `equiv.option_equiv_sum_punit` to go between `punit.{v + 1} ⊕ γ`, and `equiv.bool_equiv_punit_sum_punit ` to go between `bool` and `punit ⊕ punit` - so I guess the choice doesn't really matter.

Reply:
It probably does make sense to either use `bool` and `option` or use `punit ⊕ punit` and `punit.{v + 1} ⊕ γ` though

Reply:
since its currently a lemma specifically for my other work (which I hope to PR soon) I think I'll go with the latter two